sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -412,6 +412,16 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
           }
         }
 
+      case Or(EqualTo(l, r), And(IsNull(c1), IsNull(c2)))
```

Comment:
Assume that we have a chain of predicates combined by OR `cond1 OR cond2 OR cond3 OR ... condN`. I think we can merge `condX` and `condY` if they are `EqualTo(l, r)` and `And(IsNull(l), isNull(r))`. This is more general than the current approach.

Reply:
@cloud-fan I didn't generalize it because it would be tricky and will add complexity to code. Also it might be less common where these conditions are separated out with some other expressions in between.